plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/ProjectBuildOrderInfo.java

```diff
@@ -0,0 +1,195 @@
+		/** Injector to be used for creating instances of {@link #ProjectBuildOrderInfo()} */
+		@Inject
+		protected Injector injector;
+
+		/** Returns a new instanceof of {@link ProjectBuildOrderInfo}. No projects will be visited. */
+		@Override
+		public ProjectBuildOrderInfo get() {
+			return injector.getInstance(ProjectBuildOrderInfo.class);
+		}
+	}
+
+	/**
+	 * {@link Iterator} that iterates over {@link ProjectBuildOrderInfo#sortedProjects}.
+	 */
+	public class ProjectBuildOrderIterator implements IOrderIterator<ProjectDescription> {
+		/**
+		 * Subset of {@link #sortedProjects}: when {@link #ProjectBuildOrderInfo()} is used as an iterator, only those
+		 * projects are iterated over that are contained in this set
+		 */
+		final protected Set<String> visitProjectNames = new HashSet<>();
+		/** Iterator delegate */
+		final protected Iterator<ProjectDescription> iteratorDelegate;
+
+		ProjectBuildOrderIterator() {
+			this.iteratorDelegate = Iterables
+					.filter(sortedProjects, input -> visitProjectNames.contains(input.getName())).iterator();
+		}
+
+		@Override
+		public ProjectBuildOrderIterator visit(Collection<ProjectDescription> projectDescriptions) {
+			for (ProjectDescription prj : projectDescriptions) {
+				visitProjectNames.add(prj.getName());
+			}
+			return this;
+		}
+
+		@Override
+		public ProjectBuildOrderIterator visitAffected(List<IResourceDescription.Delta> changes) {
+			visit(getAffectedProjects(changes));
+			return this;
+		}
+
+		@Override
+		public ProjectBuildOrderIterator visitAll() {
+			visit(sortedProjects);
+			return this;
+		}
+
+		/** @return the set of projects that may contain resources that need to be rebuild given the list of changes */
+		protected Set<ProjectDescription> getAffectedProjects(List<IResourceDescription.Delta> changes) {
+			Set<String> changedProjectsNames = new HashSet<>();
+			for (IResourceDescription.Delta change : changes) {
+				XProjectManager projectManager = workspaceManager.getProjectManager(change.getUri());
+				ProjectDescription pd = projectManager.getProjectDescription();
+				changedProjectsNames.add(pd.getName());
+			}
+
+			Set<ProjectDescription> affectedProjects = new HashSet<>();
+			for (String changedProjectName : changedProjectsNames) {
+				affectedProjects.addAll(inversedDependencies.get(changedProjectName));
+			}
+
+			return affectedProjects;
+		}
+
+		@Override
+		public boolean hasNext() {
+			return iteratorDelegate.hasNext();
+		}
+
+		@Override
+		public ProjectDescription next() {
+			return iteratorDelegate.next();
+		}
+	}
+
+	/** Workspace manager */
+	@Inject
+	protected XWorkspaceManager workspaceManager;
+
+	/** Inversed set of project dependency information */
+	final protected Multimap<String, ProjectDescription> inversedDependencies = HashMultimap.create();
+	/** Build order of projects */
+	final protected List<ProjectDescription> sortedProjects = new ArrayList<>();
```

Comment:
A bit unfortunate that this list is mutable, but I see why that is the case (initialized not in ctor but in `@Inject init()`).

Reply:
That's true. We can change this in GH-1729 maybe.